### DISCUSSION TITLE
Support private maven repository in gitlab package registry

### DISCUSSION THREAD

**Discussion by simPod** [renovatebot/renovate#20090] [Request Help] [answered]
### What would you like Renovate to be able to do?

I have found docs on how to auth with artifactory https://docs.renovatebot.com/java/#custom-registry-support-and-authentication

However, the same does not seems to work for gitlab maven registry.

When bot tries to acccess private dependency it fails with 

```
DEBUG: Looking up com.domain.sth.project:somename in repository https://gitlab.domain.eu/api/v4/projects/693/packages/maven/ (repository=a/b)
DEBUG: GET https://gitlab.domain.eu/api/v4/projects/693/packages/maven/com/domain/sth/project/somename/maven-metadata.xml = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=60) (repository=a/b)
DEBUG: Content is not found for Maven url (repository=a/b)
       "url": "https://gitlab.domain.eu/api/v4/projects/693/packages/maven/com/domain/sth/project/somename/maven-metadata.xml",
       "statusCode": undefined
```

with this config json

```
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "hostRules": [
        {
            "hostType": "maven",
            "matchHost": "https://gitlab.domain.eu/api/v4/projects/693/packages/maven",
            "username": "gitlab-ci-token",
            "password": "$CI_JOB_TOKEN"
        }
    ]
}
```

### If you have any ideas on how this should be implemented, please tell us here.

Gitlab docs says to add http header `Job-Token` https://docs.gitlab.com/ee/user/packages/maven_repository/index.html#create-maven-packages-with-gitlab-cicd-using-maven

It does not seem implemented. I have scanned the renovate's repo for `Job-Token` string but found none.

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
`"password": "$CI_JOB_TOKEN"` will never work because JSON file does not do variable substitution. You need it to be a `config.js` with `password: process.env.CI_JOB_TOKEN`.

Second, the file you link to is talking about `settings.xml` field names, doesn't seem to be about http headers? Please check

**Comment by rarkins** [comment]
What you should also do is experiment with `curl https://gitlab.domain.eu/api/v4/projects/693/packages/maven/com/domain/sth/project/somename/maven-metadata.xml` until you have found which combination of headers (with token) works, then confirm it here

**Comment by simPod** [comment]
Good point on the json. I moved it to config.js as 

```js
    {
      hostType: "maven",
      matchHost: "https://gitlab.domain.eu/api/v4/projects/693/packages/maven",
      username: "gitlab-ci-token",
      password: process.env.CI_JOB_TOKEN
    }
```
However the curl authorizes successfully with code HTTP/2 200

```bash
curl -I -H "Job-Token: ${CI_JOB_TOKEN}" https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.domain.eu/api/v4/projects/693/packages/maven/com/domain/sth/project/somename/maven-metadata.xml
```

----------

The gitlab docs contain this part:

```xml
        <httpHeaders>
          <property>
            <name>Job-Token</name>
            <value>${CI_JOB_TOKEN}</value>
          </property>
        </httpHeaders>
```

We also use http header auth in app to access the private package as:

```groovy
dependencyResolutionManagement {
    repositories {
        mavenCentral()

        maven {
            name "somename"
            url "https://gitlab.domain.eu/api/v4/projects/693/packages/maven"
            credentials(HttpHeaderCredentials)
            authentication {
                header(HttpHeaderAuthentication)
            }
        }
    }
}
```

**Comment by rarkins** [comment]
The curl command doesn't directly help because you included two ways to authenticate in it. Please check how to do it with _headers_ alone, and ideally not using a non-standard header like `Job-Token` (`Authorization` is the standard).

**Comment by simPod** [comment]
Sorry my mistake.

The correct curl that works is

```
curl -I -H "Job-Token: ${CI_JOB_TOKEN}" https://gitlab.domain.eu/api/v4/projects/693/packages/maven/com/domain/sth/project/somename/maven-metadata.xml
```

What should I use instead of Job-Token?

**Comment by rarkins** [comment]
Try "Authorization" header with "Bearer $token" as the value

**Comment by simPod** [comment, marked as answer]
I did not see that coming but it worked. Since there has been some time invested into this already, I've added it into docs, see the PR https://github.com/renovatebot/renovate/pull/20089.

**Reply by viceice** [reply]
please post the correct config here too as answer

**Reply by simPod** [reply]
`token` is the key

```ts
hostRules: [
  {
    hostType: 'maven',
    matchHost: 'https://gitlab.host.com/api/v4',
    token: process.env.CI_JOB_TOKEN,
  },
];
```
